examples/lambda/wcbv-reasonable/PrelimsScript.sml

```diff
@@ -0,0 +1,491 @@
+QED
+
+Theorem nth_error_Some_lt:
+	∀n H x. nth_error n H = SOME x ⇒ n < LENGTH H
+Proof
+	Induct_on `n` >> Induct_on `H` >> rw[nth_error, EL, ADD1]
+QED
+
+Theorem nth_error_map:
+	∀n H a f. nth_error n (MAP f H) = SOME a ⇒ ∃b. nth_error n H = SOME b ∧ a = f b
+Proof
+	Induct_on `n` >> Induct_on `H` >> rw[nth_error]
+QED
+
+Theorem map_nth_error:
+	∀n H x f. nth_error n H = SOME x ⇒ nth_error n (MAP f H) = SOME (f x)
+Proof
+	Induct_on `n` >> Induct_on `H` >> rw[nth_error]
+QED
+
+Theorem nth_error_NONE_lt:
+	∀n H. nth_error n H = NONE ⇒ LENGTH H ≤ n
+Proof
+	Induct_on `n` >> Induct_on `H` >> rw[nth_error]
+QED
+
+Theorem nth_error_lt_NONE:
+	∀n H. LENGTH H ≤ n ⇒ nth_error n H = NONE
+Proof
+	Induct_on `n` >> rw[nth_error, EL, ADD1] >>
+    Induct_on `H` >> rw[nth_error, EL, ADD1] >>
+    first_x_assum drule >> rw[] >> metis_tac[nth_error, EL, ADD1]
+QED
+
+Theorem nth_error_SOME_lemma:
+	∀n H h t x.
+		nth_error n (h::t) = SOME x ⇒
+		1 <= n ⇒
+		nth_error (n-1) t = SOME x
+Proof
+	Induct_on `n` >> rw[nth_error, EL, ADD1] >>
+    Induct_on `H` >> rw[nth_error, EL, ADD1] >>
+    first_x_assum drule >> rw[] >> metis_tac[nth_error, EL, ADD1]
+QED
+
+Theorem nth_error_SOME_in_H:
+	∀n H x. nth_error n H = SOME x ⇒ MEM x H
+Proof
+	Induct_on `n` >> Induct_on `H` >> rw[nth_error]
+QED
+
+Theorem nth_error_In:
+  ∀l n x.
+    nth_error n l = SOME x ⇒ MEM x l
+Proof
+  metis_tac[nth_error_SOME_in_H]
+QED
+
+Theorem nth_error_app1:
+	∀l l' n.
+		n < LENGTH l ⇒
+    	nth_error n (l++l') = nth_error n l
+Proof
+	Induct_on `n` >> rw[nth_error, EL, ADD1] >>
+    Induct_on `l` >> rw[nth_error, EL, ADD1] >>
+    first_x_assum drule >> rw[] >> metis_tac[nth_error, EL, ADD1]
+QED
+
+Theorem nth_error_app2:
+	∀l l' n.
+		LENGTH l ≤ n ⇒
+    	nth_error n (l++l') = nth_error (n-LENGTH l) l'
+Proof
+	Induct_on `n` >> rw[nth_error, EL, ADD1] >>
+    Induct_on `l` >> rw[nth_error, EL, ADD1] >>
+    first_x_assum drule >> rw[] >> metis_tac[nth_error, EL, ADD1]
+QED
+
+Inductive Forall:
+	Forall P [] ∧
+	∀x l. P x ∧ Forall P l ⇒ Forall P (x::l)
+End
+
+Theorem Forall_map:
+	∀p f x A. Forall (λx. p (f x)) A ⇒ Forall p (MAP f A)
+Proof
+	ntac 4 strip_tac >> Induct_on `Forall` >> rw[Forall_rules]
+QED
+
+Theorem Forall_f1_imp_f2:
+	∀f1 f2 A.
+		Forall (λx. f1 x) A ⇒
+		(∀x. f1 x ⇒ f2 x) ⇒
+		Forall (λx. f2 x) A
+Proof
+	ntac 3 strip_tac >> Induct_on `Forall` >> rw[Forall_rules]
+QED
+
+Theorem Forall_forall:
+    ∀P l. Forall P l ⇔ (∀x. MEM x l ⇒ P x)
+Proof
+	Induct_on `l` >> rw[] >>
+	rw[Once Forall_cases] >> metis_tac[]
+QED
+
+Theorem Forall_MEM:
+  ∀x H P. Forall P H ∧ MEM x H ⇒ P x
+Proof
+  Induct_on `H` >> rw[]
+  >- fs[Once Forall_cases]
+  >> qpat_x_assum `Forall P (h::H)` mp_tac >>
+  rw[Once Forall_cases]
+QED
+
+Theorem in_app_or:
+  ∀l m a.
+    MEM a (l ++ m) ⇒ MEM a l ∨ MEM a m
+Proof
+  Induct_on `l` >> rw[]
+QED
+
+(* ------------------
+	   Relations
+   ------------------ *)
+
+Definition rcomp:
```

Comment:
Please use the relational composition already in `relationTheory`.  This is written with an infix capital O (`O`), or :
```
> relationTheory.O_DEF;
val it = ⊢ ∀R1 R2 x z. (R1 ∘ᵣ R2) x z ⇔ ∃y. R2 x y ∧ R1 y z: thm
```
Note though that the arguments are in the opposite order.

Reply:
Done